fe/fe-core/src/main/java/com/starrocks/sql/optimizer/Optimizer.java

```diff
@@ -341,14 +343,51 @@ private void pruneTables(OptExpression tree, TaskContext rootTaskContext, Column
             ruleRewriteIterative(tree, rootTaskContext, new MergeTwoProjectRule());
             rootTaskContext.setRequiredColumns(requiredColumns.clone());
             ruleRewriteOnlyOnce(tree, rootTaskContext, RuleSetType.PRUNE_COLUMNS);
-            context.setEnableLeftRightJoinEquivalenceDerive(true);
             ruleRewriteIterative(tree, rootTaskContext, RuleSetType.PUSH_DOWN_PREDICATE);
         }
     }
 
+    private void ruleBasedMaterializedViewRewriteMultiStages(OptExpression tree,
+                                                             TaskContext rootTaskContext,
+                                                             ColumnRefSet requiredColumns) {
+        if (!mvRewriteStrategy.mvStrategy.isMultiStages()) {
+            return;
+        }
+
+        OptimizerTraceUtil.logOptExpression("before RuleBasedMaterializedViewRewrite:\n%s", tree);
+
+        ///////////// Normalize query plan before rule based mv rewrite. /////////////
+        ruleRewriteOnlyOnce(tree, rootTaskContext, RuleSetType.PARTITION_PRUNE);
+        ruleRewriteIterative(tree, rootTaskContext, new MergeTwoProjectRule());
+        ruleRewriteIterative(tree, rootTaskContext, new MergeProjectWithChildRule());
+        CTEUtils.collectForceCteStatisticsOutsideMemo(tree, context);
```

Comment:
it seems that if we use multi stage, and these rules will be called at least twice if the plan is not rewritten by mv?

Reply:
Do we have better solution?
is the rule RewriteMultiDistinctRule conflict with mv rewrite? can we try to adjust the place of this rule, move it down to avoid it? just a idea.

Reply:
It's very diffcult to be compatible all existed rules in MV Rewrite rule.
eg:
- RewriteMultiDistinctRule
- JoinPredicatePushdown
We can split the mv rewrite into multi stage in force mode which is not set in default mode.